clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java

```diff
@@ -43,15 +43,43 @@
  */
 public class ApiVersionsResponse extends AbstractResponse {
 
+    public static final int MIN_CONSTRAINT_IBP_VERSION = 31;
```

Comment:
I'm wondering if we really need the IBP to leak into the common library. It should really only be a broker concern. Seems like the only point is so that we can continue to use the factory methods defined below from the broker code. Is that right? Could we instead move the factories to the broker?

Reply:
The tricky thing here is that if we handle the api version constraints on the broker side, it means we need to either make changes directly to the returned ApiVersionsResponse or spawn a new instance with applied constraints. That means leaking of the internal architecture of ApiVersionsResponse to the broker level and redundant conversions IMHO. The current approach makes sure the broker level logic is clean with only the necessity of passing the IBP number.